docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,440 @@
+    recovered processes are likely to have fallen behind with respect to the progressing majority.
+
+    At this point, the recovered processes are in case (g) below.
+
+- **(f)** *A new full node starts*.
+
+    The reasoning here also applies to the case when more than one full node are starting.
+    When the full node starts from scratch, it has no state (its current height is 0). Ignoring
+    statesync for the time being, the node just needs to catch up by applying past blocks one by one
+    (after verifying them).
+
+    Thus, the node is in case (g) below.
+
+- **(g)** *Advancing majority, lagging minority*
+
+    In this case, a set of full nodes, of which *n/3* or less are validators, fall behind an arbitrary
+    number of heights (e.g., temporary disconnection or network partition, memory thrashing, crashes,
+    new nodes).
+    These lagging nodes then need to catch up. They have to obtain the information they need to make
+    progress from other nodes. For each height *h*, this includes the decided block for *h*, and the
+    precommit votes also for *deciding h* (which can be extracted from the block at height *h+1*).
+    At a given height, say *h_c* a node will consider itself *caught up*, based on the (maybe out of
+    date) information it is getting from other nodes. Then, the node needs to be ready to propose
+    at height *h_c+1* which entails having received the vote extensions for *h_c*.
+
+    As the vote extensions are *not* stored in the blockchain, and it is difficult to have strong
+    guarantees on *when* a late node considers itself caught up, providing the late note with the right
+    vote extensions for the right height poses a problem.
+
+At this point, we have described and compared all cases raised in discussions leading up to this
+RFC. The list above aims at being exhaustive. The analysis of each case included above makes all of
+them converge into case (g).
+
+### Current Catch-up Mechanisms
+
+We now briefly describe the current catch-up mechanisms in the reactors concerned in Tendermint.
+
+#### Blocksync
+
+TODO
+
+#### Consensus Reactor
+
+TODO
+
+> TODO: I have considered *inserting* statesync somehow in this RFC. Do folks think this is needed?
+
+## Discussion
+
+### Solutions Proposed
+
+These are the solutions proposed in discussions leading up to this RFC.
+
+- **0.** *Vote extensions are made **best effort** in the specification*.
+
+    This is the simplest solution, considered as a way to provide vote extensions in a simple enough
+    way so that it can be part of v0.36.
+    It consists in changing the specification so as to not *require* that all precommit votes used upon
+    `PrepareProposal` contain their corresponding vote extensions. In other words, we render vote
+    extensions optional.
+    There are strong implications stemming from such a relaxation of the original specification.
```

Comment:
The way I understand your comment is that we need to provide some _flavor_ of vote extensions because app developers are waiting for them, but we don't want to introduce breakages (e.g., in the block format) that would make them not worth it. In this sense, we have Solution 3 below that fulfills these goals (and a sort of sub-optimal implementation of Solution 3 explained further below). With Solution 3, we have vote extensions with "strong" guarantees, and no breakages.
The reason I am describing Solutions 0, 1, and 2, is because they have been proposed/discussed, and we need to document them for anyone that may come, say in 1 year's time, and wants to understand why we made the choices we made.

Reply:
> Reading this document, I wonder if there's a path forward, where we don't sign vote extensions separately from the vote

This cannot be done without breaking IBC. It also means keeping extension data, I believe, as part of blocks / the chain so that we can reconstruct and verify old block signatures.

> Putting this behind a feature flag (controlled via consensus) might also mitigate some of our fears.

I think we might almost have to do this.